test/jdk/com/sun/net/httpserver/HttpServerProviderTest.java

```diff
@@ -0,0 +1,148 @@
+            case "HttpServerProviderTest$ProviderPNPC" -> testPublicNonPublicConstructor();
+            case "HttpServerProviderTest$ProviderNP" -> testNonPublic();
+            case "HttpServerProviderTest$ProviderT" -> testThrowingConstructor();
+            default -> throw new AssertionError("unexpected test case");
+        }
+    }
+
+    private void testPublic() throws Exception {
+        var n = ProviderP.class.getName();
+        assertEquals(System.getProperty(PROPERTY_KEY), n);
+
+        var p = HttpServerProvider.provider();
+        assertNull(p.createHttpServer(null, 0));
+        assertNull(p.createHttpsServer(null, 0));
+    }
+
+    private void testPublicNonPublicConstructor() {
+        var n = ProviderPNPC.class.getName();
+        assertEquals(System.getProperty(PROPERTY_KEY), n);
+
+        var e = expectThrows(ServiceConfigurationError.class, HttpServerProvider::provider);
+        assertEquals(e.getClass(), ServiceConfigurationError.class);
+        assertEquals(e.getCause().getClass(), IllegalAccessException.class);
+    }
+
+    private void testNonPublic() {
+        var n = ProviderNP.class.getName();
+        assertEquals(System.getProperty(PROPERTY_KEY), n);
+
+        var e = expectThrows(ServiceConfigurationError.class, HttpServerProvider::provider);
+        assertEquals(e.getClass(), ServiceConfigurationError.class);
+        assertEquals(e.getCause().getClass(), IllegalAccessException.class);
+    }
+
+    private void testThrowingConstructor() {
+        var cn = ProviderT.class.getName();
+        assertEquals(System.getProperty(PROPERTY_KEY), cn);
+
+        var e = expectThrows(ServiceConfigurationError.class, HttpServerProvider::provider);
+        assertEquals(e.getClass(), ServiceConfigurationError.class);
+        assertEquals(e.getCause().getClass(), InvocationTargetException.class);
+        assertEquals(e.getCause().getCause().getMessage(), "throwing constructor");
+    }
+
+    /**
+     * Test provider that is public (P)
+     */
+    public static class ProviderP extends HttpServerProvider {
+        // public default constructor
+        @Override
+        public HttpServer createHttpServer(InetSocketAddress addr, int backlog) { return null; }
+        @Override
+        public HttpsServer createHttpsServer(InetSocketAddress addr, int backlog) { return null; }
+    }
+
+    /**
+     * Test provider that is public with a non-public constructor (PNPC)
+     */
+    public static class ProviderPNPC extends HttpServerProvider {
+        ProviderPNPC() { super(); }
+        @Override
+        public HttpServer createHttpServer(InetSocketAddress addr, int backlog) { return null; }
+        @Override
+        public HttpsServer createHttpsServer(InetSocketAddress addr, int backlog) { return null; }
+    }
+
+    /**
+     * Test provider that is not public (NP)
+     */
+    static class ProviderNP extends HttpServerProvider {
+        // package-private default constructor
+        @Override
```

Comment:
The class is pp here, while the default constructor ( as generated by the compiler  ) is public. So the comment needs a little adjustment, or better still just add an explicit no-args public constructor.

Reply:
The default constructor has the same access modifier as the class (or is pp respectively), as per JLS 8.8.9. This being said, it's better to spell out the access control here, so I'll add explicit constructors to all test provider classes.